atomic_reactor/util.py

```diff
@@ -626,3 +627,54 @@ def get_credentials(self, docker_registry):
         except KeyError:
             logger.warn('%s not found in .dockercfg', docker_registry)
             return {}
+
+
+def get_manifest_digests(image, insecure=False, dockercfg_path=None,
+                         versions=('v1', 'v2')):
+    """Return manifest digest for image.
+
+    :param image: ImageName, the remote image to inspect
+    :param insecure: bool, when True registry's cert is not verified
+    :param dockercfg_path: str, dirname of .dockercfg location
+    :param versions: tuple, which manifest schema versions to fetch digest
+
+    :return: dict, versions mapped to their digest
+    """
+    auth = None
+    if dockercfg_path:
+        dockercfg = Dockercfg(dockercfg_path).get_credentials(image.registry)
+
+        username = dockercfg.get('username')
+        password = dockercfg.get('password')
+        if username and password:
+            auth = requests.auth.HTTPBasicAuth(username, password)
+
+    context = '/'.join([x for x in [image.namespace, image.repo] if x])
+    tag = image.tag or 'latest'
+    url = 'https://{}/v2/{}/manifests/{}'.format(image.registry, context, tag)
+
+    digests = {}
+    for version in versions:
+        media_type = ('application/vnd.docker.distribution.manifest.{}+json'
+                      .format(version))
+        headers = {'Accept': media_type}
+        response = requests.head(url, verify=not insecure, headers=headers,
+                                 auth=auth)
+        response.raise_for_status()
+
+        # Only compare prefix as response may use +prettyjws suffix
+        # which is the case for signed manifest
+        response_h_prefix = response.headers['Content-Type'].rsplit('+', 1)[0]
+        request_h_prefix = media_type.rsplit('+', 1)[0]
+        if response_h_prefix != request_h_prefix:
+            logger.debug('request headers: %s', headers)
+            logger.debug('response headers: %s', response.headers)
+            raise RuntimeError('Unable to fetch digest for {}'
+                               .format(media_type))
```

Comment:
I'm not sure we want this to error out here. Instead, we could either just set the digest for the requested version, or leave it unset and require that calling code wanting e.g. v2 must first try v2 and fall back to v1.

The reason raising an error might be bad is that someone might want to run this code using a Docker version earlier than 1.10, or against a v2 registry that does not support schema 2 (e.g. docker-distribution < 2.3.0), and we still want that set-up to work.

In that situation, we want e.g. Koji to use whatever digest the registry is storing the manifest for.